### PR TITLE
Add catalog read-only sharing to all organizations

### PR DESCRIPTION
**Pull request by dataclouder**
Add types and functions to handle read-only catalog sharing (Address [Issue 987](https://github.com/vmware/terraform-provider-vcd/issues/987))

* `AdminCatalog.SetReadOnlyAccessControl`
* `Catalog.SetReadOnlyAccessControl`
* `AdminCatalog.IsSharedReadOnly`
* `Catalog.IsSharedReadOnly`

Note: setting read-only access control for other organizations is the result of two API calls: the first one is an access control request with all parameters set to their zero values, and the second one is a `/action/publish` (not the same as `/action/publishToExternalOrganizations`, but will set the catalog's `IsPublished` status)

### CODE REVIEW

**Comment by juanfont** [comment]
I confirm this is working :)

Any change it will be merged soon?

File: govcd/access_control.go [hunk 1]
```diff
@@ -456,3 +456,144 @@ func checkSanityVdcControlAccess(vdc *Vdc) error {
...
+	err = cat.Refresh()
+	if err != nil {
+		return false, err
+	}
+	return cat.AdminCatalog.IsPublished, nil
+}
+
+// publish publishes a catalog read-only access control to all organizations
+// This operation is usually the second step for a read-only sharing to all Orgs
+func (cat *AdminCatalog) publish(isPublished bool) error {
```
Comment (adezxc) [inline govcd/access_control.go:556]: nit, but could you move the functions so the context is a bit lighter to understand? catalog.publish adminCatalog.publish catalog.IsSharedReadOnly adminCatalog.IsSharedReadOnly etc.

Comment (adezxc) [inline govcd/access_control.go:556]: nit, but could you move the functions so the context is a bit lighter to understand? catalog.publish adminCatalog.publish catalog.IsSharedReadOnly adminCatalog.IsSharedReadOnly etc.

File: govcd/access_control.go [hunk 2]
```diff
@@ -456,3 +456,144 @@ func checkSanityVdcControlAccess(vdc *Vdc) error {
...
+	if cat.AdminCatalog == nil {
+		return fmt.Errorf("cannot publish catalog, Object is empty")
+	}
+
+	catalogUrl := cat.AdminCatalog.HREF
+	if catalogUrl == "nil" || catalogUrl == "" {
```
Comment (adezxc) [inline govcd/access_control.go:562]: Just curious as I didn't find it in the code, but do we set `cat.AdminCatalog.HREF` to `"nil"` somewhere? I saw the if statement in other places though

File: govcd/catalog_test.go [hunk 3]
```diff
@@ -1307,6 +1307,161 @@ func (vcd *TestVCD) Test_CatalogAccessAsOrgUsers(check *C) {
...
+	time.Sleep(timeElapsedToAvailability)
+	adminCatalog2FromOrg, err = vcdClient2.Client.GetAdminCatalogByName(org1.AdminOrg.Name, catalogName)
+	check.Assert(err, NotNil)
+	check.Assert(adminCatalog2FromOrg, IsNil)
+
+	isSharedReadOnly, err = adminCatalog1.IsSharedReadOnly()
+	check.Assert(err, IsNil)
+	check.Assert(isSharedReadOnly, Equals, false)
+
+	timeout = 30 * time.Second
```
Comment (adezxc) [inline govcd/catalog_test.go:1453]: What would happen if the deletion takes more than 30 seconds and there's a failure happening every time inside of the `for` loop? Would it be hard to check for errors after successful/unsuccessful deletion?

File: .changes/v2.20.0/559-features.md [hunk 4]
```diff
@@ -0,0 +1 @@
+* Added methods `SetReadOnlyAccessControl` and `IsSharedReadOnly` `Catalog` and `AdminCatalog`, to handle read-only catalog sharing [GH-559]
```
Comment (adezxc) [inline .changes/v2.20.0/559-features.md:1]:
```suggestion
* Added methods `SetReadOnlyAccessControl` and `IsSharedReadOnly` for `Catalog` and `AdminCatalog`, to handle read-only catalog sharing [GH-559]
```

**Review by adezxc — APPROVED** [review]
Tested locally, worked great! Just some general questions

File: govcd/catalog_test.go (hunk 3 above)
Comment (dataclouder) [reply govcd/catalog_test.go:1453]: Right. An assertion is missing after the loop. Thanks

File: govcd/access_control.go (hunk 1 above)
Comment (dataclouder) [reply govcd/access_control.go:541]: done

File: .changes/v2.20.0/559-features.md (hunk 4 above)
Comment (dataclouder) [reply .changes/v2.20.0/559-features.md]: Fixed. Incidentally, I moved the changelog entries from `v2.20.0` (when they were created!) to the current `v2.22.0`

**Review by Didainius — APPROVED** [review]

File: govcd/access_control.go [hunk 5]
```diff
@@ -456,3 +456,145 @@ func checkSanityVdcControlAccess(vdc *Vdc) error {
...
+		return false, nil
+	}
+	err = cat.Refresh()
+	if err != nil {
+		return false, err
+	}
+	return cat.AdminCatalog.IsPublished, nil
+}
+
+// publish publishes a catalog read-only access control to all organizations
```
Comment (lvirbalas) [inline govcd/access_control.go:511]: Nit, separator needed for two sentences:

```suggestion
// publish publishes a catalog read-only access control to all organizations.
```

File: types/v56/types.go [hunk 6]
```diff
@@ -1134,6 +1134,14 @@ type PublishExternalCatalogParams struct {
 	PreserveIdentityInfoFlag *bool  `xml:"PreserveIdentityInfoFlag,omitempty"` // True includes BIOS UUIDs and MAC addresses in the downloaded OVF package. If false, those information will be excluded.
 }
 
+// PublishCatalogParams represents the configuration parameters of a catalog published to other orgs
```
Comment (lvirbalas) [inline types/v56/types.go:1137]: Nit, separator needed for two sentences:

```suggestion
// PublishCatalogParams represents the configuration parameters of a catalog published to other orgs.
```

File: govcd/access_control.go [hunk 7]
```diff
@@ -456,3 +456,145 @@ func checkSanityVdcControlAccess(vdc *Vdc) error {
...
+	}
+	err = publishCatalog(cat.client, catalogUrl, tenantContext, publishParameters)
+	if err != nil {
+		return err
+	}
+
+	return cat.Refresh()
+}
+
+// publish publishes a catalog read-only access control to all organizations
```
Comment (lvirbalas) [inline govcd/access_control.go:539]: Nit:

```suggestion
// publish publishes a catalog read-only access control to all organizations.
```

**Review by lvirbalas — APPROVED** [review]
Just some polishing asks :)

File: govcd/access_control.go (hunk 7 above)
Comment (dataclouder) [reply govcd/access_control.go]: fixed

**Review by adambarreiro — APPROVED** [review]
LGTM

**Pull request merged**